agent/agentserver/server_test.go

```diff
@@ -276,13 +277,82 @@ func TestReadinessCheckHandler(t *testing.T) {
 	}
 }
 
+func TestReadinessCheckHandlerCache(t *testing.T) {
+	for _, tc := range []struct {
+		desc              string
+		firstCallErr      error
+		readinessCacheTTL time.Duration
+		waitInvalidation  bool
+	}{
+		{
+			desc:              "call 1 succeeds, so second call automatically succeeds",
+			firstCallErr:      nil,
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+		},
+		{
+			desc:              "call 1 fails, so second call performs checks",
+			firstCallErr:      errors.New("test-error"),
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+		},
+		{
+			desc:              "call 1 succeeds, but cache becomes invalid, so second call performs checks",
+			firstCallErr:      nil,
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  true,
+		},
+		{
+			desc:              "call 1 succeeds, but caching is disabled, so second call performs checks",
+			firstCallErr:      nil,
+			readinessCacheTTL: 0,
+			waitInvalidation:  false,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			require := require.New(t)
+			mocks, cleanup := newServerMocks(t)
+			defer cleanup()
+
+			mocks.sched.EXPECT().Probe().Return(tc.firstCallErr).Times(1)
+			mocks.tags.EXPECT().CheckReadiness().Return(tc.firstCallErr).Times(1)
+			mocks.ac.EXPECT().CheckReadiness().Return(tc.firstCallErr).Times(1)
+			if tc.firstCallErr != nil || tc.waitInvalidation || tc.readinessCacheTTL == 0 {
+				mocks.sched.EXPECT().Probe().Return(nil).Times(1)
+				mocks.tags.EXPECT().CheckReadiness().Return(nil).Times(1)
+				mocks.ac.EXPECT().CheckReadiness().Return(nil).Times(1)
+			}
+
+			s, addr := mocks.startServer(Config{readinessCacheTTL: tc.readinessCacheTTL})
+			r, err := httputil.Get(fmt.Sprintf("http://%s/readiness", addr))
+			if tc.firstCallErr == nil {
+				require.Nil(err)
+				respBody, _ := ioutil.ReadAll(r.Body)
+				require.Equal("OK", string(respBody))
+			} else {
+				require.Error(err)
+			}
+
+			if tc.waitInvalidation {
+				// To avoid using time.Sleep, we can rollback the server's lastReady variable to simulate cache invalidation.
+				s.lastReady = s.lastReady.Add(-1 * tc.readinessCacheTTL)
```

Comment:
this is an optional thing here since it's not super complicated but if you have more complicated clock logic use a clock lib with a mock https://pkg.go.dev/github.com/andres-erbsen/clock

Reply:
Noted. Will keep in mind for the future.